File: Week3/Task_3.1P/3.1 - Answer/CounterTask/Clock.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CounterTask
{
    public class Clock
    {
        private Counter _hours;
        private Counter _minutes;
        private Counter _seconds;
        public Clock()
        {
            _hours = new Counter("");
            _minutes = new Counter("");
            _seconds = new Counter("");
        }
        
        public int Hours
        {
            get => _hours.Tick;
        }

        public int Minutes
        {
            get => _minutes.Tick;
        }

        public int Seconds
        {
            get => _seconds.Tick;
        }

        public void Tick()
        {
            if (_seconds.Tick < 59)
            {
                _seconds.Increment();
            }
            else if (_minutes.Tick < 59)
            {
                _seconds.Reset();
                _minutes.Increment();
            }
            else if (_hours.Tick < 23)
            {
                _seconds.Reset();
                _minutes.Reset();
                _hours.Increment();
            }
            else
            {
                _hours.Reset();
                _minutes.Reset();
                _seconds.Reset();
            }
        }
        public void Reset()
        {
            _hours.Reset();
            _minutes.Reset();
            _seconds.Reset();
        }
        public string PrintTime()
        {
            string currentTime = _hours.Tick.ToString("D2") + ":" + _minutes.Tick.ToString("D2") + ":" + _seconds.Tick.ToString("D2");
            Console.WriteLine(currentTime);

            return currentTime;
        }
        public void StartClock(int seconds)
        {
            for (int i = 0; i < seconds; i++)
            {
                Thread.Sleep(1000);
                Tick();
                PrintTime();
            }
        }
    }
}

File: Week11/Clock.py
import time
from Counter import Counter

class Clock:
    def __init__(self):
        self._hours = Counter("Hours")
        self._minutes = Counter("Minutes")
        self._seconds = Counter("Seconds")

    @property
    def hours(self):
        return self._hours.tick

    @property
    def minutes(self):
        return self._minutes.tick

    @property
    def seconds(self):
        return self._seconds.tick

    def tick(self):
        if self._seconds.tick < 59:
            self._seconds.increment()
        elif self._minutes.tick < 59:
            self._seconds.reset()
            self._minutes.increment()
        elif self._hours.tick < 23:
            self._seconds.reset()
            self._minutes.reset()
            self._hours.increment()
        else:
            self._hours.reset()
            self._minutes.reset()
            self._seconds.reset()

    def reset(self):
        self._hours.reset()
        self._minutes.reset()
        self._seconds.reset()

    def print_time(self):
        current_time = f"{self._hours.tick:02}:{self._minutes.tick:02}:{self._seconds.tick:02}"
        print(current_time)
        return current_time

    def start_clock(self, seconds):
        for _ in range(seconds):
            time.sleep(1)
            self.tick()
            self.print_time()